feat(profile) Добавление функционала для получения ачивок пользователя

V3rgi9ly opened the pull request:
-(Controller) Создал класс AchievementController с GET запросом /api/achievements для получения ачивок пользователя
-(Config) Создал класс AchievementConfig для получения   информации об описании ачивок из настроек файла application.yaml
-(apllication.yaml) Добавил словарь в файл apllication.yml с описанием ачивок
-(Service) Добавлен метод getProfileAchievements в класс AchievementService, который отправляет в контроллер массив ачивок профиля
-(Repository) Добавил метод в класс AchievementRepository для нахождения ачивок пользователя по его айди профилю
-(Enum) Добавил новые типы ачивок в класс Enum
-(DTO) Создал класс ProfileAchievementsDto для передачи данных ачивок из AchievementService в контроллер
-(http test) Добавил http тесты для проверки всех успешных и неуспешных сценариев работы get запроса

V3rgi9ly committed:
feat(profile) Добавление функционала для получения ачивок пользователя

-(Controller) Создал класс AchievementController с GET запросом /api/achievements для получения ачивок пользователя
-(Config) Создал класс AchievementConfig для получения   информации об описании ачивок из настроек файла application.yaml
-(apllication.yaml) Добавил словарь в файл apllication.yml с описанием ачивок
-(Service) Добавлен метод getProfileAchievements в класс AchievementService, который отправляет в контроллер массив ачивок профиля
-(Repository) Добавил метод в класс AchievementRepository для нахождения ачивок пользователя по его айди профилю
-(Enum) Добавил новые типы ачивок в класс Enum
-(DTO) Создал класс ProfileAchievementsDto для передачи данных ачивок из AchievementService в контроллер
-(http test) Добавил http тесты для проверки всех успешных и неуспешных сценариев работы get запроса

V3rgi9ly committed:
Merge branch 'dev' into feature-getting-achievements-36

# Conflicts:
#	src/main/java/com/itmentorcommunityplatform/profileservice/domain/type/AchievementType.java
#	src/main/java/com/itmentorcommunityplatform/profileservice/dto/ProfileAchievementsDto.java
#	src/main/java/com/itmentorcommunityplatform/profileservice/repository/AchievementRepository.java

V3rgi9ly committed:
fix(profile) Исправление конфликтов между веткой dev и feature-getting-achievements-36

-(Controller) Изменил тип передаваемого объекта с ProfileAchievementsDto на AchievementsDto
-(apllication.yaml) Изменил тип ачивок
-(Service) Изменил тип передаваемого объекта в контроллер с ProfileAchievementsDto на AchievementsDto
-(Repository) Добавил метод в класс AchievementRepository для нахождения ачивок пользователя по его айди профилю
-(Enum) Изменил тип ачивок
-(DTO) Создал класс AchievementsDto для передачи данных ачивок из AchievementService в контроллер

DarkRubin commented on src/main/java/com/itmentorcommunityplatform/profileservice/config/AchievementConfig.java:
```diff
@@ -0,0 +1,23 @@
+@ConfigurationProperties(prefix = "app")
+public class AchievementConfig {
+
+    private  Map<String, String> achievements;
```
Ключем должен быть класс `AchievementType` чтобы нельзя было передать несуществующее значение, были автоподсказки от IDE конфиге, и не нужно было делать вызов `toString` для получения значения

DarkRubin commented on src/main/java/com/itmentorcommunityplatform/profileservice/config/AchievementConfig.java:
```diff
@@ -0,0 +1,23 @@
+    private  Map<String, String> achievements;
+
+
+    public Map<String, String>  getAchievements(){
```
В других местах в проекте для создания геттеров и сеттеров используются `lombok` аннотации, придерживайся существующего в проекте стиля написания кода

DarkRubin commented on httpRequests/achievements-tests.http:
```diff
@@ -1,13 +1,39 @@
+}
+
+### Успешный сценарий. Получение ачивок пользователя
+GET http://localhost:8083/api/achievements
```
Пропустил это в прошлом PR. Ручные http тесты должны быть отдельно от исходных файлов кода проекта, в уже существующей в данном случае для этого отдельной директории. Также непонятно, что здесь делает запрос к совершенно другому микросервису

DarkRubin commented on src/main/java/com/itmentorcommunityplatform/profileservice/dto/AchievementsDto.java:
```diff
@@ -0,0 +1,9 @@
+
+import com.itmentorcommunityplatform.profileservice.domain.type.AchievementType;
+
+public record AchievementsDto(AchievementType type,
```
Во множественном числе называют только классы констант (например `HttpHeaders`) или когда класс действительно являет собой обертку для хранения множества значений одного типа. В данном случае здесь поля с информацией про конкретное достижение, так что название должно быть в едином числе

DarkRubin commented on src/main/java/com/itmentorcommunityplatform/profileservice/dto/AchievementsDto.java:
```diff
@@ -0,0 +1,9 @@
+import com.itmentorcommunityplatform.profileservice.domain.type.AchievementType;
+
+public record AchievementsDto(AchievementType type,
+                              Long earned_timestamp,
```
Название поля нарушает стандартные Java конвенции нейминга полей и переменных. Если в JSON нужно отдавать поле с иным названием для этого можно использовать аннотацию `@JsonProperty`, можешь посмотреть примеры использования в других дто в проекте

DarkRubin commented on src/main/java/com/itmentorcommunityplatform/profileservice/repository/AchievementRepository.java:
```diff
@@ -5,10 +5,11 @@
 
     boolean existsByProfileIdAndAchievementType(Long profileId, AchievementType achievementType);
-
+    Optional<List<Achievement>> findAchievemetsByProfileId(Long id);
```
Не понял зачем результат обернут в `Optional`. Даже если записи с таким `profileId` не будет в БД то просто вернется пустой список

DarkRubin commented on src/main/java/com/itmentorcommunityplatform/profileservice/repository/AchievementRepository.java:
```diff
@@ -5,10 +5,11 @@
     boolean existsByProfileIdAndAchievementType(Long profileId, AchievementType achievementType);
-
+    Optional<List<Achievement>> findAchievemetsByProfileId(Long id);
     List<Achievement> findAllByProfileIdAndPubliclyVisibleTrue(Long profileId);
```
Также сохраняй между сигнатурами методов отступ в одну пусту строку, даже в интерфейсах. Так намного легче смотреть что-то, особенно когда методов стает больше

DarkRubin commented on src/main/java/com/itmentorcommunityplatform/profileservice/controller/AchievementController.java:
```diff
@@ -0,0 +1,30 @@
+    private final AchievementService achievementService;
+
+    @GetMapping
+    public ResponseEntity<List<AchievementsDto>> getProfileAchievements(
```
Не добавляю отдельным пунктом в требования к задаче, но покрывай все ново добавленные эндпоинты Swagger документацией, если он уже добавлен в проект

DarkRubin commented on src/main/java/com/itmentorcommunityplatform/profileservice/service/AchievementService.java:
```diff
@@ -34,6 +46,22 @@ public void recheckAndAwardAchievements(ProjectCreatedEvent event) {
 
+    public List<AchievementsDto> getProfileAchievements(Long telegramUserId){
+
+        AchievementsDto achievementsDto;
```
В джава нет необходимости выносить объявление переменной, отдельно от её инициализации в цикле, для этого есть автоматические оптимизации. Но читать код, когда переменная объявляется в том же месте, где и используется, чуть проще

DarkRubin commented on src/main/java/com/itmentorcommunityplatform/profileservice/service/AchievementService.java:
```diff
@@ -34,6 +46,22 @@ public void recheckAndAwardAchievements(ProjectCreatedEvent event) {
+        AchievementsDto achievementsDto;
+        List<AchievementsDto> profileAchievementsDtoList=new ArrayList<>();
+        Optional<Profile> profile= profileRepository.findByTelegramUserId(telegramUserId);
+        List<Achievement> achievement=achievementRepository.findAchievemetsByProfileId(profile.get().getId()).orElseThrow(()->new ResponseStatusException(HttpStatus.NOT_FOUND, "Achievements not found"));
```
Тут сразу ряд замечаний к двум строкам выше:
* `Optional` переменная используется без проверки на наличие содержимого
* Если содержимого действительно нет запрос падает с 500 и не понятной пользователю ошибкой `No value present`
*  Название переменной `achievement` вместо логичного `achievements`
* Здесь в одной строке 200 символов и выполняется 4+ действия, вводи промежуточные переменные и делай отступ при вызове функциональных методов (по типу `orElseThrow`)
* Про то, что этот кейс можно обрабатывать проще, уже написал в репозитории

DarkRubin commented on src/main/java/com/itmentorcommunityplatform/profileservice/service/AchievementService.java:
```diff
@@ -34,6 +46,22 @@ public void recheckAndAwardAchievements(ProjectCreatedEvent event) {
+        Optional<Profile> profile= profileRepository.findByTelegramUserId(telegramUserId);
+        List<Achievement> achievement=achievementRepository.findAchievemetsByProfileId(profile.get().getId()).orElseThrow(()->new ResponseStatusException(HttpStatus.NOT_FOUND, "Achievements not found"));
+
+        for (Achievement achiev:achievement){
```
То же самое, что и до этого: отсутствие форматирования, и не самый понятный нейминг переменных. Перед комитом/пушем файлов проходись ещё раз свежим взглядом по всем измененным файлам, и не игнорируй предупреждения от IDE

DarkRubin commented on src/main/java/com/itmentorcommunityplatform/profileservice/service/AchievementService.java:
```diff
@@ -34,6 +46,22 @@ public void recheckAndAwardAchievements(ProjectCreatedEvent event) {
+        for (Achievement achiev:achievement){
+           String description= achievementConfig.getAchievements().get(achiev.getAchievementType().toString());
+            achievementsDto=new AchievementsDto(achiev.getAchievementType(), achiev.getEarnedTimestamp(), description, achiev.isPubliclyVisible());
+           profileAchievementsDtoList.add(achievementsDto);
```
Логика эндпоинта не совпадает описанием в требованиях, он должен возвращать все существующие достижения в системе, и если они есть у текущего пользователя — устанавливать значение для `earned_timestamp`. Сейчас у тебя возвращается просто список полученных достижений студента

DarkRubin commented on src/main/resources/application.yaml:
```diff
@@ -28,3 +28,13 @@ management:
           - health
           - prometheus
 
+app:
```
Вместо `app` лучше использовать название этого сервиса:`profile-service`

V3rgi9ly committed:
fix(profile) Исправление ошибок и замечаний

-(Controller) Добавил Swagger аннотацию @GetProfileAchievements для метода getProfileAchievements
-(apllication.yaml) Изменил название словаря app на profile-service
-(Service) Изменил логику работы метода getProfileAchievements. Теперь данные метод передает все 5 ачивок, учитывая есть ли у профиля достижения
-(Repository) Изменил отступы методов в классе AchievementRepository. В методе findAchievemetsByProfileId удалил обертку класса Optional на список класса Achievement
-(Swagger) Добавил класс GetProfileAchievements в котором воспользовался инструментом Swagger для документирования работы класса контроллера AchievementController
-(DTO) Изменил класс AchievementsDto с record на публичный. В полях earnedTimestamp и publiclyVisible назначил аннотацию @JsonProperty для сериализации нейминга в json формат
-(http tests)Создал и перенес https тесты в файл achievements-tests.http

V3rgi9ly committed:
fix(profile) Исправление ошибок и замечаний

-(Controller) Добавил Swagger аннотацию @GetProfileAchievements для метода getProfileAchievements
-(apllication.yaml) Изменил название словаря app на profile-service
-(Service) Изменил логику работы метода getProfileAchievements. Теперь данные метод передает все 5 ачивок, учитывая есть ли у профиля достижения
-(Repository) Изменил отступы методов в классе AchievementRepository. В методе findAchievemetsByProfileId удалил обертку класса Optional на список класса Achievement
-(Swagger) Добавил класс GetProfileAchievements в котором воспользовался инструментом Swagger для документирования работы класса контроллера AchievementController
-(DTO) Изменил класс AchievementsDto с record на публичный. В полях earnedTimestamp и publiclyVisible назначил аннотацию @JsonProperty для сериализации нейминга в json формат
-(http tests)Создал и перенес https тесты в файл achievements-tests.http

DarkRubin commented on httpRequests/createProjectConsumerRequests.http:
```diff
@@ -0,0 +1,17 @@
+### Успешный сценарий. Добавление проекта в бд
+POST http://localhost:8084//api/projects/project
```
Этого запроса не должно быть вообще в этом репозитории

DarkRubin commented on src/main/java/com/itmentorcommunityplatform/profileservice/service/AchievementService.java:
```diff
@@ -34,6 +45,36 @@ public void recheckAndAwardAchievements(ProjectCreatedEvent event) {
+
+        }
+
+        return listAchievementsProfile;
```
Рабочий вариант, но можно было сделать чуть оптимальнее, без вложенных циклов

```java
public List<AchievementDto> getProfileAchievements(Long telegramUserId) {
    Map<AchievementType, String> achievementsDescriptions = achievementConfig.getAchievements();

    Long profileId = profileRepository.findByTelegramUserId(telegramUserId)
            .map(Profile::getId)
            .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Profile not found"));

    Map<AchievementType, AchievementDto> profileAchievements = achievementRepository.findAchievemetsByProfileId(profileId)
            .stream()
            .map(achievement -> new AchievementDto(
                    achievement.getAchievementType(),
                    achievement.getEarnedTimestamp(),
                    achievementsDescriptions.get(achievement.getAchievementType()),
                    achievement.isPubliclyVisible()
            ))
            .collect(Collectors.toMap(AchievementDto::getType, Function.identity()));

    achievementsDescriptions.forEach((type, name) -> {
        AchievementDto achievementDto = new AchievementDto(type, 0L, name, true);
        profileAchievements.putIfAbsent(type, achievementDto);
    });

    return new ArrayList<>(profileAchievements.values());
}
```

DarkRubin commented on httpRequests/achievements-tests.http:
```diff
@@ -0,0 +1,23 @@
+Accept: application/json
+X-Telegram-User-Id: 987654321
+
+### Неуспешный сценарий. Отсутствует обязательный загаловок X-Telegram-User-Id:
```
Добавь, также как в остальных .http файлах запросов,  HTTP коды ответов: (200 OK), (404 Not found)

V3rgi9ly committed:
fix(profile) Исправление ошибок и замечаний

-(Service) Изменил логику работы метода getProfileAchievements, согласно описанным рекомендациям
-(http tests)Удалил http файл createProjectConsumerRequests.http так как данный запрос не должен находится в данном репозитории. Добавил в файл achievements-tests.http в заголовка коды с результатами запросов

V3rgi9ly committed:
fix(profile) Исправление ошибок и замечаний

-(Service) Изменил логику работы метода getProfileAchievements, согласно описанным рекомендациям
-(http tests)Удалил http файл createProjectConsumerRequests.http так как данный запрос не должен находится в данном репозитории. Добавил в файл achievements-tests.http в заголовке коды с результатами запросов
